src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLConv.scala

```diff
@@ -0,0 +1,245 @@
+        filter_out_channels, filter_in_channels, filter_height, filter_width)
+      */
+      val filter_descriptor = get_descriptor(right, "filter")
+
+      // create convolution descriptor
+      val conv_descriptor = new CUDNN_CONV_DESCRIPTOR(NEW_STRUCT(manifest[CUDNN_CONV_DESCRIPTOR], "cudnnConvolutionDescriptor_t").x)
+      CUDNN_CREATE_CONV_DESCRIPTOR(conv_descriptor)
+      CUDNN_SET_CONV_2D_DESCRIPTOR(conv_descriptor, INT(padding(0)), INT(padding(1)), INT(strides(0)), INT(strides(1)), INT(dilation(0)), INT(dilation(1)),
+        mode, datatype)
+
+      // find output tensor shape
+      var output_batchsize = 0
+      var output_height = 0
+      var output_width = 0
+      var output_channels = 0
+      CUDNN_GET_CONV_2D_FWD_OUTPUT_DIM(conv_descriptor, input_descriptor, filter_descriptor, 
+        INT(output_batchsize), INT(output_channels), INT(output_height), INT(output_width))
+
+      // create output descriptor
+      val output_descriptor = new CUDNN_TENSOR_DESCRIPTOR(NEW_STRUCT(manifest[CUDNN_CONV_DESCRIPTOR], "cudnnConvolutionDescriptor_t").x)
+      CUDNN_CREATE_TENSOR_DESCRIPTOR(output_descriptor)
+      CUDNN_SET_TENSOR_4D_DESCRIPTOR(output_descriptor, layout, datatype, 
+        INT(output_batchsize), INT(output_channels), INT(output_height), INT(output_width))
+      // val output_descriptor = get_descriptor(s, "tensor")  // TODO: use get_descriptor for output_descriptor
+
+      // allocate output tensor
+      val output_size = output_batchsize * output_height * output_width * output_channels
+      val output = gpu_array(output_size, manifest[Float], myNCCLRank)
+      
+      // find convolution algorithm
+      var res_count = 0
+      val res = new CUDNN_CONV_FWD_ALG_PERF(NEW_STRUCT(manifest[CUDNN_CONV_FWD_ALG_PERF], "cudnnConvolutionFwdAlgoPerf_t").x)
+      CUDNN_FIND_CONV_FWD_ALG(myCUDNNComm, input_descriptor, filter_descriptor, conv_descriptor, output_descriptor, INT(1), INT(res_count), res)
+      // val conv_algo = READ_FIELD(manifest[CUDNN_CONV_FWD_ALG_PERF], manifest[CUDNN_CONV_FWD_ALGO], res.x, "algo")
+      val convAlgoRep = readField[Manifest[CUDNN_CONV_FWD_ALG_PERF], Manifest[CUDNN_CONV_FWD_ALGO]](Wrap[Manifest[CUDNN_CONV_FWD_ALG_PERF]](res.x), "algo")
+      val convAlgo = TOP(Unwrap(convAlgoRep), manifest[CUDNN_CONV_FWD_ALGO])
+
+      // allocate convolution workspace
+      var workspace_bytes = 0
+      CUDNN_GET_CONV_FWD_WORKSPACE_SZ(myCUDNNComm, input_descriptor, filter_descriptor, conv_descriptor, output_descriptor,
+        convAlgo, SIZE_T(workspace_bytes))
+      // var d_workspace = CUDA_MALLOC(workspace_bytes, manifest[FLOAT])
+      val d_workspace = gpu_array(workspace_bytes, manifest[Float], myNCCLRank)
+
+      // convolution
+      CUDNN_CONV_FWD(myCUDNNComm, VAR(INT(alpha)), input_descriptor, leftTensor, filter_descriptor, rightTensor, conv_descriptor, convAlgo, d_workspace, 
+        SIZE_T(workspace_bytes), VAR(INT(beta)), output_descriptor, sourceTensor)
```

Comment:
IIRC the `CUDNN_CONV_FWD` should operate on `Array`s instead of `Tensor`s. So I think the use of `leftTensor`, `rightTensor` and `sourceTensor` here is probably incorrect. 
The `leftTensor` should be `transform(left_operand)`. The `rightTensor` should be `transform(right_operand)`, and the `sourceTensor` should be a new GPUArray.

Reply:
fixed in 33dc2e5